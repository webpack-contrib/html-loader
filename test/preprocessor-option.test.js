import Handlebars from "handlebars";
import posthtml from "posthtml";
import posthtmlWebp from "posthtml-webp";

import {
  compile,
  execute,
  getCompiler,
  getErrors,
  getModuleSource,
  getWarnings,
  readAsset,
} from "./helpers";

describe("'process' option", () => {
  it('should work with the "preprocessor" option', async () => {
    const compiler = getCompiler("preprocessor.hbs", {
      preprocessor: (content, loaderContext) => {
        expect(typeof content).toBe("string");
        expect(loaderContext).toBeDefined();

        let result;

        try {
          result = Handlebars.compile(content)({
            firstname: "Alexander",
            lastname: "Krasnoyarov",
          });
        } catch (error) {
          loaderContext.emitError(error);

          return content;
        }

        return result;
      },
    });
    const stats = await compile(compiler);

    expect(getModuleSource("./preprocessor.hbs", stats)).toMatchSnapshot(
      "module"
    );
    expect(
      execute(readAsset("main.bundle.js", compiler, stats))
    ).toMatchSnapshot("result");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should work with Async "preprocessor" Function option', async () => {
    const compiler = getCompiler("preprocessor.hbs", {
      preprocessor: async (content, loaderContext) => {
        await expect(typeof content).toBe("string");
        await expect(loaderContext).toBeDefined();

        let result;

        try {
          result = await Handlebars.compile(content)({
            firstname: "Alexander",
            lastname: "Krasnoyarov",
          });
        } catch (error) {
          await loaderContext.emitError(error);

          return content;
        }

        return result;
      },
    });
    const stats = await compile(compiler);

    expect(getModuleSource("./preprocessor.hbs", stats)).toMatchSnapshot(
      "module"
    );
    expect(
      execute(readAsset("main.bundle.js", compiler, stats))
    ).toMatchSnapshot("result");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should work with the "preprocessor" option #2', async () => {
    const plugin = posthtmlWebp();
    const compiler = getCompiler("posthtml.html", {
      preprocessor: (content, loaderContext) => {
        expect(typeof content).toBe("string");
        expect(loaderContext).toBeDefined();

        let result;

        try {
          result = posthtml().use(plugin).process(content, { sync: true });
        } catch (error) {
          loaderContext.emitError(error);

          return content;
        }

        return result.html;
      },
    });
    const stats = await compile(compiler);

    expect(getModuleSource("./posthtml.html", stats)).toMatchSnapshot("module");
    expect(
      execute(readAsset("main.bundle.js", compiler, stats))
    ).toMatchSnapshot("result");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });
  it('should work with the Async "preprocessor" Function option #2', async () => {
    const plugin = posthtmlWebp();
    const compiler = getCompiler("posthtml.html", {
      preprocessor: async (content, loaderContext) => {
        await expect(typeof content).toBe("string");
        await expect(loaderContext).toBeDefined();

        let result;

        try {
          result = await posthtml()
            .use(plugin)
            .process(content, { sync: true });
        } catch (error) {
          loaderContext.emitError(error);

          return content;
        }

        return result.html;
      },
    });
    const stats = await compile(compiler);

    expect(getModuleSource("./posthtml.html", stats)).toMatchSnapshot("module");
    expect(
      execute(readAsset("main.bundle.js", compiler, stats))
    ).toMatchSnapshot("result");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });
});
