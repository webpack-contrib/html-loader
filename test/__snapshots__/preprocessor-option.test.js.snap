// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'process' option should work with Async "preprocessor" Function option: errors 1`] = `[]`;

exports[`'process' option should work with Async "preprocessor" Function option: module 1`] = `
"// Imports
import ___HTML_LOADER_GET_SOURCE_FROM_IMPORT___ from "../../src/runtime/getUrl.js";
var ___HTML_LOADER_IMPORT_0___ = new URL("./image.png", import.meta.url);
// Module
var ___HTML_LOADER_REPLACEMENT_0___ = ___HTML_LOADER_GET_SOURCE_FROM_IMPORT___(___HTML_LOADER_IMPORT_0___);
var code = "<div>\\n  <p>Alexander Krasnoyarov</p>\\n  <img src=\\"" + ___HTML_LOADER_REPLACEMENT_0___ + "\\" alt=\\"alt\\" />\\n<div>\\n";
// Exports
export default code;"
`;

exports[`'process' option should work with Async "preprocessor" Function option: result 1`] = `
"<div>
  <p>Alexander Krasnoyarov</p>
  <img src="replaced_file_protocol_/webpack/public/path/image.png" alt="alt" />
<div>
"
`;

exports[`'process' option should work with Async "preprocessor" Function option: warnings 1`] = `[]`;

exports[`'process' option should work with the "preprocessor" option #2: errors 1`] = `[]`;

exports[`'process' option should work with the "preprocessor" option #2: module 1`] = `
"// Imports
import ___HTML_LOADER_GET_SOURCE_FROM_IMPORT___ from "../../src/runtime/getUrl.js";
var ___HTML_LOADER_IMPORT_0___ = new URL("./image.png.webp", import.meta.url);
var ___HTML_LOADER_IMPORT_1___ = new URL("./image.png", import.meta.url);
// Module
var ___HTML_LOADER_REPLACEMENT_0___ = ___HTML_LOADER_GET_SOURCE_FROM_IMPORT___(___HTML_LOADER_IMPORT_0___);
var ___HTML_LOADER_REPLACEMENT_1___ = ___HTML_LOADER_GET_SOURCE_FROM_IMPORT___(___HTML_LOADER_IMPORT_1___);
var code = "<picture><source type=\\"image/webp\\" srcset=\\"" + ___HTML_LOADER_REPLACEMENT_0___ + "\\"><img src=\\"" + ___HTML_LOADER_REPLACEMENT_1___ + "\\"></picture>\\n";
// Exports
export default code;"
`;

exports[`'process' option should work with the "preprocessor" option #2: result 1`] = `
"<picture><source type="image/webp" srcset="replaced_file_protocol_/webpack/public/path/image.png.webp"><img src="replaced_file_protocol_/webpack/public/path/image.png"></picture>
"
`;

exports[`'process' option should work with the "preprocessor" option #2: warnings 1`] = `[]`;

exports[`'process' option should work with the "preprocessor" option: errors 1`] = `[]`;

exports[`'process' option should work with the "preprocessor" option: module 1`] = `
"// Imports
import ___HTML_LOADER_GET_SOURCE_FROM_IMPORT___ from "../../src/runtime/getUrl.js";
var ___HTML_LOADER_IMPORT_0___ = new URL("./image.png", import.meta.url);
// Module
var ___HTML_LOADER_REPLACEMENT_0___ = ___HTML_LOADER_GET_SOURCE_FROM_IMPORT___(___HTML_LOADER_IMPORT_0___);
var code = "<div>\\n  <p>Alexander Krasnoyarov</p>\\n  <img src=\\"" + ___HTML_LOADER_REPLACEMENT_0___ + "\\" alt=\\"alt\\" />\\n<div>\\n";
// Exports
export default code;"
`;

exports[`'process' option should work with the "preprocessor" option: result 1`] = `
"<div>
  <p>Alexander Krasnoyarov</p>
  <img src="replaced_file_protocol_/webpack/public/path/image.png" alt="alt" />
<div>
"
`;

exports[`'process' option should work with the "preprocessor" option: warnings 1`] = `[]`;

exports[`'process' option should work with the Async "preprocessor" Function option #2: errors 1`] = `[]`;

exports[`'process' option should work with the Async "preprocessor" Function option #2: module 1`] = `
"// Imports
import ___HTML_LOADER_GET_SOURCE_FROM_IMPORT___ from "../../src/runtime/getUrl.js";
var ___HTML_LOADER_IMPORT_0___ = new URL("./image.png.webp", import.meta.url);
var ___HTML_LOADER_IMPORT_1___ = new URL("./image.png", import.meta.url);
// Module
var ___HTML_LOADER_REPLACEMENT_0___ = ___HTML_LOADER_GET_SOURCE_FROM_IMPORT___(___HTML_LOADER_IMPORT_0___);
var ___HTML_LOADER_REPLACEMENT_1___ = ___HTML_LOADER_GET_SOURCE_FROM_IMPORT___(___HTML_LOADER_IMPORT_1___);
var code = "<picture><source type=\\"image/webp\\" srcset=\\"" + ___HTML_LOADER_REPLACEMENT_0___ + "\\"><img src=\\"" + ___HTML_LOADER_REPLACEMENT_1___ + "\\"></picture>\\n";
// Exports
export default code;"
`;

exports[`'process' option should work with the Async "preprocessor" Function option #2: result 1`] = `
"<picture><source type="image/webp" srcset="replaced_file_protocol_/webpack/public/path/image.png.webp"><img src="replaced_file_protocol_/webpack/public/path/image.png"></picture>
"
`;

exports[`'process' option should work with the Async "preprocessor" Function option #2: warnings 1`] = `[]`;
